program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
+                self.invaldate_following_entries(
+                    cmp::min(num_skipped_entries, NUM_ENTRIES),
+                    prev_entry,
+                );
+                self.max_update_time_gap = update_time;
+            }
+        }
+        Ok(())
+    }
+}
+
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an Tick Tracker that has NUM_ENTRIES entries
+/// each tracking the last tick before every GRANUALITY seconds.
+pub struct TickTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    prices:         [SignedTrackerRunningSum; NUM_ENTRIES], /* price running time
+                                                             * weighted sums */
+    confidences:    [UnsignedTrackerRunningSum; NUM_ENTRIES], /* confidence running
+                                                               * time
+                                                               * weighted sums */
+    entry_validity: [u8; NUM_ENTRIES], //entry validity
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    ///invalidates current_entry, and increments self.current_entry num_entries times
+    fn invaldate_following_entries(&mut self, num_entries: usize, current_entry: usize) {
+        let mut entry = current_entry;
+        for _ in 0..num_entries {
+            self.entry_validity[entry] = 0;
+            entry = Self::get_next_entry(entry);
+        }
+    }
+}
+
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+    for TickTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        Ok(())
+    }
 
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        let prev_entry = Self::time_to_entry(prev_time)?;
+        let current_entry = Self::time_to_entry(current_time)?;
+        let num_skipped_entries = current_entry - prev_entry;
+        match num_skipped_entries {
+            0 => {
+                self.prices[current_entry] = current_price;
+                self.confidences[current_entry] = current_conf;
+            }
+            1 => {
+                self.prices[prev_entry] = prev_price;
+                self.confidences[prev_entry] = prev_conf;
+                self.entry_validity[prev_entry] =
+                    if Self::get_time_to_entry_end(prev_time) >= THRESHOLD {
+                        0
+                    } else {
+                        1
+                    };
+                self.prices[current_entry] = current_price;
+                self.confidences[current_entry] = current_conf;
+            }
+            _ => {
```

Comment:
You can probably do the same behavior for 1 and _ in this match statement.

Reply:
- updates to entries, 1 only updates one entry. I can do something similar to SMA trackers, and have a helper that does 1 on an entery and call it twice in _. Does that sound better?